File: cli/app_keys.ts
/**
 * Application API Keys management module.
 */

import { ask, askNumber, pressEnter, clearScreen, confirm } from "./utils/prompt.ts";
import { section, table, success, error, info, apiKeyBox, colors, color } from "./utils/display.ts";
import { createApiKey, listApiKeys, revokeApiKey, deleteApiKey } from "../db/api_keys.ts";
import { migrate } from "../db/migrate.ts";

export async function manageAppKeys(): Promise<void> {
  // Ensure migrations are run
  await migrate();

  while (true) {
    clearScreen();
    section("🎫 API Keys de la Aplicación");

    const keys = listApiKeys();

    if (keys.length === 0) {
      info("No hay API Keys creadas.");
      console.log();
      console.log("  1. Crear nueva API Key");
      console.log("  0. Volver");
      console.log();

      const choice = await askNumber("> ");
      if (choice === 1) {
        await createNewKey();
      } else {
        return;
      }
      continue;
    }

    // Build table
    const headers = ["ID", "Prefix", "Nombre", "Último uso", "Estado", "Usos"];
    const rows = keys.map((k) => [
      String(k.id),
      k.key_prefix,
      k.name || "-",
      k.last_used_at?.slice(0, 16) || "-",
      k.is_active
        ? color("✓ Activa", colors.green)
        : color("✗ Revocada", colors.red),
      String(k.usage_count),
    ]);

    table(headers, rows, [5, 13, 16, 18, 12, 8]);

    console.log();
    console.log("  1. Crear nueva API Key");
    console.log("  2. Revocar API Key");
    console.log("  3. Eliminar API Key");
    console.log("  0. Volver");
    console.log();

    const choice = await askNumber("> ");

    switch (choice) {
      case 1:
        await createNewKey();
        break;
      case 2:
        await revokeKey();
        break;
      case 3:
        await deleteKey();
        break;
      case 0:
      default:
        return;
    }
  }
}

async function createNewKey(): Promise<void> {
  console.log();
  const name = await ask("Nombre para la nueva API Key (opcional): ");

  try {
    const { key, record } = await createApiKey(name.trim() || undefined);

    apiKeyBox(key);

    console.log("  Detalles:");
    console.log(`    ID:      ${record.id}`);
    console.log(`    Nombre:  ${record.name || "(sin nombre)"}`);
    console.log(`    Creada:  ${record.created_at}`);
    console.log();

    info("Usa esta key en el header 'Authorization: Bearer <key>'");
    info("o en el header 'x-api-key: <key>'");

  } catch (err) {
    error(`Error creando API Key: ${err}`);
  }

  await pressEnter();
}

async function revokeKey(): Promise<void> {
  console.log();
  const id = await askNumber("ID de la API Key a revocar: ");

  if (!id) {
    info("Operación cancelada");
    await pressEnter();
    return;
  }

  const confirmed = await confirm(`¿Revocar API Key #${id}?`);

  if (!confirmed) {
    info("Operación cancelada");
    await pressEnter();
    return;
  }

  const result = revokeApiKey(id);

  if (result) {
    success(`API Key #${id} revocada`);
    info("La key ya no funcionará para autenticarse");
  } else {
    error(`API Key #${id} no encontrada`);
  }

  await pressEnter();
}

async function deleteKey(): Promise<void> {
  console.log();
  const id = await askNumber("ID de la API Key a eliminar: ");

  if (!id) {
    info("Operación cancelada");
    await pressEnter();
    return;
  }

  const confirmed = await confirm(`¿ELIMINAR PERMANENTEMENTE API Key #${id}?`);

  if (!confirmed) {
    info("Operación cancelada");
    await pressEnter();
    return;
  }

  const result = deleteApiKey(id);

  if (result) {
    success(`API Key #${id} eliminada permanentemente`);
  } else {
    error(`API Key #${id} no encontrada`);
  }

  await pressEnter();
}

File: internal/cli/appkeys.go
package cli

import (
	"context"
	"fmt"
)

// ManageAppKeys implements menu option 3: manage application API keys.
func ManageAppKeys(app *App) {
	ctx := context.Background()

	for {
		ClearScreen()
		Section("🎫 API Keys de la Aplicación")

		keys, err := app.Keys.List(ctx)
		if err != nil {
			Error(fmt.Sprintf("Error leyendo API Keys: %v", err))
			PressEnter("")
			return
		}

		if len(keys) == 0 {
			Info("No hay API Keys creadas.")
			fmt.Println()
			fmt.Println("  1. Crear nueva API Key")
			fmt.Println("  0. Volver")
			fmt.Println()
			n, _, _ := AskNumber("> ")
			if n == 1 {
				createAppKey(app, ctx)
			} else {
				return
			}
			continue
		}

		// Table: ID, Prefix, Nombre, Último uso, Estado, Usos
		headers := []string{"ID", "Prefix", "Nombre", "Último uso", "Estado", "Usos"}
		rows := make([][]string, len(keys))
		for i, k := range keys {
			name := "-"
			if k.Name.Valid {
				name = k.Name.String
			}
			lastUsed := "-"
			if k.LastUsedAt.Valid && len(k.LastUsedAt.String) >= 16 {
				lastUsed = k.LastUsedAt.String[:16]
			}
			var estado string
			if k.IsActive == 1 {
				estado = Color(Green, "✓ Activa")
			} else {
				estado = Color(Red, "✗ Revocada")
			}
			rows[i] = []string{
				fmt.Sprintf("%d", k.ID),
				k.KeyPrefix,
				name,
				lastUsed,
				estado,
				fmt.Sprintf("%d", k.UsageCount),
			}
		}
		Table(headers, rows, []int{4, 12, 15, 17, 12, 6})

		fmt.Println()
		fmt.Println("  1. Crear nueva API Key")
		fmt.Println("  2. Revocar API Key")
		fmt.Println("  3. Eliminar API Key")
		fmt.Println("  0. Volver")
		fmt.Println()

		n, _, _ := AskNumber("> ")
		switch n {
		case 1:
			createAppKey(app, ctx)
		case 2:
			revokeAppKey(app, ctx)
		case 3:
			deleteAppKey(app, ctx)
		default:
			return
		}
	}
}

func createAppKey(app *App, ctx context.Context) {
	fmt.Println()
	name, _ := Ask("Nombre para la nueva API Key (opcional): ")

	plain, rec, err := app.Keys.Create(ctx, name)
	if err != nil {
		Error(fmt.Sprintf("Error creando API Key: %v", err))
		PressEnter("")
		return
	}

	ApiKeyBox(plain)

	fmt.Println("  Detalles:")
	fmt.Printf("    ID:      %d\n", rec.ID)
	n := "(sin nombre)"
	if rec.Name.Valid {
		n = rec.Name.String
	}
	fmt.Printf("    Nombre:  %s\n", n)
	fmt.Printf("    Creada:  %s\n", rec.CreatedAt)
	fmt.Println()
	Info("Usa esta key en el header 'Authorization: Bearer <key>'")
	Info("o en el header 'x-api-key: <key>'")

	PressEnter("")
}

func revokeAppKey(app *App, ctx context.Context) {
	fmt.Println()
	id, ok, err := AskNumber("ID de la API Key a revocar: ")
	if err != nil || !ok || id == 0 {
		Info("Operación cancelada")
		PressEnter("")
		return
	}

	confirmed, _ := Confirm(fmt.Sprintf("¿Revocar API Key #%d?", id), false)
	if !confirmed {
		Info("Operación cancelada")
		PressEnter("")
		return
	}

	found, err := app.Keys.Revoke(ctx, int64(id))
	if err != nil {
		Error(fmt.Sprintf("Error: %v", err))
	} else if found {
		Success(fmt.Sprintf("API Key #%d revocada", id))
		Info("La key ya no funcionará para autenticarse")
	} else {
		Error(fmt.Sprintf("API Key #%d no encontrada", id))
	}
	PressEnter("")
}

func deleteAppKey(app *App, ctx context.Context) {
	fmt.Println()
	id, ok, err := AskNumber("ID de la API Key a eliminar: ")
	if err != nil || !ok || id == 0 {
		Info("Operación cancelada")
		PressEnter("")
		return
	}

	confirmed, _ := Confirm(fmt.Sprintf("¿ELIMINAR PERMANENTEMENTE API Key #%d?", id), false)
	if !confirmed {
		Info("Operación cancelada")
		PressEnter("")
		return
	}

	found, err := app.Keys.Delete(ctx, int64(id))
	if err != nil {
		Error(fmt.Sprintf("Error: %v", err))
	} else if found {
		Success(fmt.Sprintf("API Key #%d eliminada permanentemente", id))
	} else {
		Error(fmt.Sprintf("API Key #%d no encontrada", id))
	}
	PressEnter("")
}
